elasticsearch/api.go

```diff
@@ -0,0 +1,238 @@
+	if err != nil {
+		log.ErrorC("failed to create request for call to elastic", err, logData)
+		return nil, 0, err
+	}
+
+	resp, err := api.client.Do(ctx, req)
+	if err != nil {
+		log.ErrorC("failed to call elastic", err, logData)
+		return nil, 0, err
+	}
+	defer resp.Body.Close()
+
+	logData["http_code"] = resp.StatusCode
+
+	jsonBody, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		log.ErrorC("failed to read response body from call to elastic", err, logData)
+		return nil, resp.StatusCode, err
+	}
+	logData["json_body"] = string(jsonBody)
+	logData["status_code"] = resp.StatusCode
+
+	if resp.StatusCode < http.StatusOK || resp.StatusCode >= 300 {
+		log.ErrorC("failed", ErrorUnexpectedStatusCode, logData)
+		return nil, resp.StatusCode, ErrorUnexpectedStatusCode
+	}
+
+	return jsonBody, resp.StatusCode, nil
+}
+
+// Body represents the request body to elasticsearch
+type Body struct {
+	From      int        `json:"from"`
+	Size      int        `json:"size"`
+	Highlight *Highlight `json:"highlight,omitempty"`
+	Query     Query      `json:"query"`
+	Sort      []Scores   `json:"sort"`
+}
+
+// Highlight represents parts of the fields that matched
+type Highlight struct {
+	PreTags  []string          `json:"pre_tags,omitempty"`
+	PostTags []string          `json:"post_tags,omitempty"`
+	Fields   map[string]Object `json:"fields,omitempty"`
+	Order    string            `json:"score,omitempty"`
+}
+
+// Object represents an empty object (as expected by elasticsearch)
+type Object struct{}
+
+// Query represents the request query details
+type Query struct {
+	Bool Bool `json:"bool"`
+}
+
+// Bool represents the desirable goals for query
+type Bool struct {
+	Must   []Match `json:"must,omitempty"`
+	Should []Match `json:"should,omitempty"`
+}
+
+// Match represents the fields that the term should or must match too in query
+type Match struct {
+	Match map[string]string `json:"match,omitempty"`
+}
+
+// Scores represents a list of scoring, e.g. scoring on relevance, but can add in secondary
+// score such as alphabetical order if relevance is the same for two search results
+type Scores struct {
+	Score Score `json:"_score"`
+}
+
+// Score contains the ordering of the score (ascending ot descending)
+type Score struct {
+	Order string `json:"order"`
+}
+
+func buildSearchQuery(term string, limit, offset int) *Body {
+	var object Object
+	highlight := make(map[string]Object)
+
+	highlight["label"] = object
+	highlight["code"] = object
+
+	label := make(map[string]string)
+	code := make(map[string]string)
+	label["label"] = term
+	code["code"] = term
+
+	labelMatch := Match{
+		Match: label,
+	}
+
+	codeMatch := Match{
+		Match: code,
+	}
+
+	scores := Scores{
+		Score: Score{
+			Order: "desc",
+		},
+	}
+
+	listOfScores := []Scores{}
+	listOfScores = append(listOfScores, scores)
+
+	query := &Body{
+		From: offset,
+		Size: limit,
+		Highlight: &Highlight{
+			PreTags:  []string{"\001S"},
+			PostTags: []string{"\001E"},
```

Comment:
these two lines are inconsistent with other uses of the tags (which use the `"\u0001S"` Unicode forms) - can they be standardised?